### PR TITLE
Enhancement/generator initialization

### LINKED ISSUE #65049: Setting a random seed or passing a generator to Modules for deterministic sampling.
## 🚀 Feature

I would like to obtain deterministic/reproducible weight initialization when working with say Con1d, while limiting the deterministic state to the weight [default] initialization part only and retaining the random state for the rest of the torch modules.

## Motivation

With this functionality the models can be in initialized in a reproducible fashion while still allowing randomness in shuffling and thereby the optimization path taken by the optimizer. This is particularly interesting when studying the properties of different optimizer and architectures.  

## Pitch

Following, non-functional, code expresses the idea I am trying to achieve

```
gen1 = torch.Generator()
gen1.manual_seed(123)
conv1 = nn.Conv1d(16, 33, 3, stride=2)
nn.init.kaiming_uniform_(conv.weight, a=0, mode='fan_in', nonlinearity='leaky_relu', generator=gen1)


gen2 = torch.Generator()
gen2.manual_seed(123)
conv2 = nn.Conv1d(16, 33, 3, stride=2)
nn.init.kaiming_uniform_(conv2.weight, a=0, mode='fan_in', nonlinearity='leaky_relu', generator=gen2)

assert conv2.weight == conv1.weight
```
Expected outcome: True

## Alternatives
Until now I have to use following work around which at times is tedious since I have to recreate the different initializers

```
gen1 = torch.Generator()
gen1.manual_seed(123)
conv1 = nn.Conv1d(16, 33, 3, stride=2)
bound = # some calculation to compute the bound for He Initialization
x = (-bound - bound) * torch.rand((33, 16, 3), generator=gen1) + bound
conv1.weight.data = x

gen2 = torch.Generator()
gen2.manual_seed(123)
conv2 = nn.Conv1d(16, 33, 3, stride=2)
bound = # some calculation to compute the bound for He Initialization
x = (-bound - bound) * torch.rand((33, 16, 3), generator=gen2) + bound
conv2.weight.data = x

assert torch.eq(conv2.weight.data, conv1.weight.data) 
```

### PR DESCRIPTION
**Pull request by khatrishubham88**
Fixes #65049

Need to run tests on CI

### CODE REVIEW

**Comment by khatrishubham88** [comment]
The CLA has been signed.

**Comment by khatrishubham88** [comment]
@albanD Ideas on why this JIT test is failing?
It says that it cannot find type annotation torch.Generator which seems very strange.

File: torch/nn/init.py [hunk 1]
```diff
@@ -119,23 +119,24 @@ def calculate_gain(nonlinearity, param=None):
...
     Examples:
         >>> w = torch.empty(3, 5)
         >>> nn.init.uniform_(w)
     """
-    return _no_grad_uniform_(tensor, a, b)
+    return _no_grad_uniform_(tensor, a, b, generator=generator)
 
 
-def normal_(tensor: Tensor, mean: float = 0., std: float = 1.) -> Tensor:
+def normal_(tensor: Tensor, mean: float = 0., std: float = 1., generator: torch.Generator = None) -> Tensor:
```
Comment (albanD) [inline torch/nn/init.py:139]: The doc below should be updated right?

**Review by albanD — COMMENTED** [review]
Thanks for the PR, that looks quite good.

Could you also add tests for uniform_, normal_ and trunc_normal_ ?

**Comment by khatrishubham88** [comment]
> Thanks for the PR, that looks quite good.
> 
> Could you also add tests for uniform_, normal_ and trunc_normal_ ?

Sure thing. I will write a test for those. I missed it since it wasn't clear if those are really needed, I don't find any prior test for them.

File: test/test_nn.py [hunk 2]
```diff
@@ -11764,6 +11764,20 @@ def test_xavier_uniform(self):
                 bounds = expected_std * math.sqrt(3)
                 assert self._is_uniform(input_tensor, -bounds, bounds)
 
+    @unittest.skipIf(not TEST_SCIPY, "Scipy not found.")
```
[the remark this answers is not in the document]
Comment (khatrishubham88) [reply test/test_nn.py:11767]: The problem with merging them with above is that those test have a bit of a different semantic than what the new test is doing. And I personally think that merging them could make things a bit fuzzy.

Additionally, there is an added complication of random float sampling happening in the previous tests when gain parameter is used and enforcing reproducibility there would require quite a few changes in test. For example, addition of python rng in `_random_float` method of test, addition of rng in `_create_random_nd_tensor` method. I wasn't sure if such changes would be welcomed. So if you give green on that I can merge them.

File: torch/nn/init.py (hunk 1 above)
Comment (khatrishubham88) [reply torch/nn/init.py:139]: Yes the doc below is updated.

File: test/test_nn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/test_nn.py:11767]: Ok, we can have them separate then! Thanks for the detailed explanation.

**Comment by khatrishubham88** [comment]
> > Thanks for the PR, that looks quite good.
> > Could you also add tests for uniform_, normal_ and trunc_normal_ ?
> 
> Sure thing. I will write a test for those.

@albanD completed in 47e5d3297f102e6d0527c3ca29e75e61586e4284

**Comment by khatrishubham88** [comment]
@albanD @jbschlosser Do you think it makes sense to finish off https://github.com/pytorch/pytorch/issues/65049#issuecomment-920431499 too in this?

**Comment by albanD** [comment]
> @albanD @jbschlosser Do you think it makes sense to finish off #65049 (comment) too in here?

I don't think we should do that here no.
The change to core nn.Module was more controversial than this one. So we should get this one in first :D 

Master was a bit broken yesterday, can you rebase now to make sure all CI is green?

**Review by albanD — APPROVED** [review]
Test update looks good!

**Comment by khatrishubham88** [comment]
> > @albanD @jbschlosser Do you think it makes sense to finish off #65049 (comment) too in here?
> 
> I don't think we should do that here no.
> The change to core nn.Module was more controversial than this one. So we should get this one in first :D
> 
> Master was a bit broken yesterday, can you rebase now to make sure all CI is green?

Sure thing, will be here in a bit.

**Comment by jbschlosser** [comment]
> > > @albanD @jbschlosser Do you think it makes sense to finish off #65049 (comment) too in here?
> > 
> > 
> > I don't think we should do that here no.
> > The change to core nn.Module was more controversial than this one. So we should get this one in first :D
> > Master was a bit broken yesterday, can you rebase now to make sure all CI is green?
> 
> Sure thing, will be here in a bit.
> 
> Should I bring up the proposed changes somewhere for discussion?

Agree with @albanD; we don't want to update all module signatures at this time. Feel free to open a new issue requesting this for discussion.

Also incidentally it looks like something may have gone wrong with the rebase - a bunch of unrelated commits / code changes made it in :(

**Comment by khatrishubham88** [comment]
> > > > @albanD @jbschlosser Do you think it makes sense to finish off #65049 (comment) too in here?
> > > 
> > > 
> > > I don't think we should do that here no.
> > > The change to core nn.Module was more controversial than this one. So we should get this one in first :D
> > > Master was a bit broken yesterday, can you rebase now to make sure all CI is green?
> > 
> > 
> > Sure thing, will be here in a bit.
> > Should I bring up the proposed changes somewhere for discussion?
> 
> Agree with @albanD; we don't want to update all module signatures at this time. Feel free to open a new issue requesting this for discussion.

Sure thing I would do that.

> Also incidentally it looks like something may have gone wrong with the rebase - a bunch of unrelated commits / code changes made it in :(

Ammn, when I forked the repo from that time to the time I rebased bunch of changes happened. So when I rebase my branch on master these changes are bound to appear. Maybe I didn't understand you correctly?

Also ideas on why this jit thing is appearing? It should detect the torch.Generator, but it isn't.

**Comment by albanD** [comment]
Hey!
I fixed the branch (sorry for the force push in your fork).
Let's see if CI is green now.

**Comment by khatrishubham88** [comment]
> Hey!
> I fixed the branch (sorry for the force push in your fork).
> Let's see if CI is green now.

Might I ask what did you change here? Did you rebase the branch on master again?

**Comment by albanD** [comment]
I removed the merge commit at the top (doing a hard reset to the last commit with some changes) and then rebased on top of master with (git rebase master).

**Comment by albanD** [comment]
Let me take a look at the jit issues.

**Comment by khatrishubham88** [comment]
> I removed the merge commit at the top (doing a hard reset to the last commit with some changes) and then rebased on top of master with (git rebase master).

Interesting, I wonder what went wrong when I did the same....

**Comment by albanD** [comment]
cc @eellison This is blocked because jit is registering these ops as special ops. But it does't recognise the Generator type there.
What is the way forward to be able to do this change?

@khatrishubham88 Sorry, this is going to take a bit longer than expected while we fix this jit issue. Nothing for you to do though.

**Review by eellison — COMMENTED** [review]
cc @gmagogsfm, need to add support for generator, should not be too bad I dont think

**Comment by khatrishubham88** [comment]
> cc @gmagogsfm, need to add support for generator, should not be too bad I dont think

@eellison @gmagogsfm any updates?

**Comment by albanD** [comment]
@eellison most of the random API in pytorch use this already though. Are all of them not support in TS either?

**Comment by khatrishubham88** [comment]
> @eellison most of the random API in pytorch use this already though. Are all of them not support in TS either?

Could you point me to the module where the support needs to be added? Maybe I can do it..

**Comment by khatrishubham88** [comment]
@albanD @eellison @gmagogsfm @jbschlosser  Any updates?
Recommended starting point so I can fix it myself?

**Comment by eellison** [comment]
@khatrishubham88 sorry I pinged someone who was on vacation. @albanD you're right it is supported there just hasn't been the right binding for it on python land.  

I think the fixes are:
- Need to bind `GeneratorType` [here]( andhttps://github.com/pytorch/pytorch/blob/20143bf07f425c773d3e378459634348e01f790a/torch/csrc/jit/python/python_ir.cpp#L860)
- Need to add a case for `if ann is torch.Generator`: [here](https://github.com/pytorch/pytorch/blob/master/torch/jit/annotations.py#L367)

from there I would ctrl-F for StreamObjType - I suspect that should pop up all of the cases you'll need to add to handle Generator here

**Comment by khatrishubham88** [comment]
> * Need to bind `GeneratorType` here

@eellison Is there a missing link here?

**Comment by eellison** [comment]
@khatrishubham88 yes sorry,  https://github.com/pytorch/pytorch/blob/20143bf07f425c773d3e378459634348e01f790a/torch/csrc/jit/python/python_ir.cpp but also yea just searching should show the places you need to add handling:

https://github.com/pytorch/pytorch/search?q=StreamObjType
